### 10. NUMPY – BINARY OPERATORS

NumPy-யில் **binary operators** பயன்படுத்தி bitwise operations செய்யலாம்.
Bitwise operations binary (இரும) data-களில் வேலை செய்யும், அதாவது
bit-அளவிலான data-ஐ நேரடியாக மாற்றும். இவை data manipulation மற்றும்
bit-level operations-ஐ எளிதாக்கும்.

#### 10.1. numpy.bitwise_and

**bitwise_and** operator இரண்டு arrays-இன் bitwise AND operation-ஐ return
செய்கிறது. இது ஒவ்வொரு bit-இல் உள்ள values-ஐ எடுத்து AND operation செய்கிறது,
அதாவது, இரண்டு bit-களும் 1 என்றால் மட்டுமே முடிவு 1 ஆக இருக்கும்.



In [4]:
import numpy as np

# Bitwise AND operation
a = np.array([0b1100])
b = np.array([0b1010])
result = np.bitwise_and(a, b)
print("Bitwise AND:", result)

Bitwise AND: [8]





-   **a** = 0b1100 (binary) = 12 (decimal)
-   **b** = 0b1010 (binary) = 10 (decimal)
-   Bitwise AND operation (1100 AND 1010) = 1000 (binary) = 8 (decimal)

**bitwise_and** operation AND logic பயன்படுத்தி இரண்டு arrays-இன்
corresponding bit-களை compare செய்து, முடிவுகளை return செய்கிறது.

#### 10.2. numpy.bitwise_or

**bitwise_or** operator இரண்டு arrays-இன் bitwise OR operation-ஐ return
செய்கிறது. இது OR operation-ஐ ஒவ்வொரு bit-ஐ வைத்து செய்கிறது, அதாவது,
bit-களில் எதாவது ஒரு value 1 என்றால், முடிவாக 1 return ஆகும்.

In [5]:
# Bitwise OR operation
a = np.array([0b1100])
b = np.array([0b1010])
result = np.bitwise_or(a, b)
print("Bitwise OR:", result)

Bitwise OR: [14]



-   **a** = 0b1100 (binary) = 12 (decimal)
-   **b** = 0b1010 (binary) = 10 (decimal)
-   Bitwise OR operation (1100 OR 1010) = 1110 (binary) = 14 (decimal)

**bitwise_or** operation OR logic-ஐ பயன்படுத்தி bit-level-ல் values-ஐ
compare செய்து, final result-ஐ return செய்கிறது.

#### 10.3. numpy.invert()

**invert()** function-ஐ பயன்படுத்தி array-இல் உள்ள ஒவ்வொரு bit-ஐ எதிர்மறையாக
மாற்றலாம். இதன் பொருள், bit 1 இருந்தால் 0 ஆகவும், 0 இருந்தால் 1 ஆகவும் மாற்றம்
செய்யும்.


In [6]:
# Bitwise NOT operation
a = np.array([0b1100], dtype=np.uint8)
result = np.invert(a)
print("Bitwise NOT:", result)

Bitwise NOT: [243]



-   **a** = 0b1100 (binary) = 12 (decimal)
-   **invert()** operation bit-களில் inversion செய்கிறது, அதாவது 1100
    becomes 11110011 (for an 8-bit integer) = 243 (decimal).

**invert()** operation bit-level inversion-ஐ எளிதாக செய்து, data-ஐ
logically toggle செய்ய உதவுகிறது.

#### 10.4. numpy.left_shift

**left_shift** operator-ஐ பயன்படுத்தி array-இன் bits-ஐ இடது பக்கம் (left
side) நகர்த்தலாம். இது bit values-ஐ shift செய்து, அதற்கு வலது பக்கம் புதிய 0s
சேர்க்கும்.

In [7]:
# Left shift operation
a = np.array([0b0010])
result = np.left_shift(a, 2)
print("Left Shift:", result)

Left Shift: [8]



-   **a** = 0b0010 (binary) = 2 (decimal)
-   Left shift operation shifts the bits 2 positions to the left: 0010
    becomes 1000 (binary) = 8 (decimal).

**left_shift** operation bit-level-ல் values-ஐ இடது பக்கம் நகர்த்தி, binary
representation-ஐ மாற்ற உதவுகிறது.

#### 10.5. numpy.right_shift

**right_shift** operator-ஐ பயன்படுத்தி array-இன் bits-ஐ வலது பக்கம் (right
side) நகர்த்தலாம். இது bit values-ஐ shift செய்து, அதற்கு இடது பக்கம் 0s
சேர்க்கும்.

In [8]:
# Right shift operation
a = np.array([0b1000])
result = np.right_shift(a, 2)
print("Right Shift:", result)


Right Shift: [2]


#### Explanation:

-   **a** = 0b1000 (binary) = 8 (decimal)
-   Right shift operation shifts the bits 2 positions to the right: 1000
    becomes 0010 (binary) = 2 (decimal).

**right_shift** operation bit-level-ல் values-ஐ வலது பக்கம் நகர்த்தி, binary
data-ஐ சுருக்க உதவுகிறது.

இந்த binary operators data-ஐ bit-level-ல் manipulate செய்யவும், binary data
handling ஐ memory-efficient-ஆகவும் செய்ய உதவுகின்றன.